### PR TITLE
Replace PyCrypto with cryptography.

### PR DESCRIPTION
**Pull request by xoviat**
I know the cryptodome route is easier, but cryptography is the way to go for the long run because it is supported by the python software foundation. Feel free to give feedback and suggestions!

### CODE REVIEW

**Comment by mpdavis** [comment]
Unfortunately, there are many people that rely on the PyCrypto support in environments where cryptography is not supported, such as Google App Engine.

That said, with the recent work around custom algorithms would make it pretty easy to add a implementation based on cryptography that ships with python-jose that could easily be registered and used.  I would even be open to conversations about setting it as the default, but PyCrypto would still need to be supported.

**Comment by xoviat** [comment]
> Unfortunately, there are many people that rely on the PyCrypto support in environments where cryptography is not supported, such as Google App Engine.

This is really Google's narrative; that there is "investment" and "community support" for these old libraries. I understand that from your perspective, PyCrypto support is necessary, but the only reason why is Google, frankly. And I'm not suggesting that this is an unreasonable perspective to have, considering that GAE standard is considerably cheaper than GAE flexible.

What I think will be satisfactory is provide a backend that minimally emulates cryptography using pycrypto. I'm currently down to 3 test failures on pypy only (which tests the pycrypto backend), so that should be completed today.

**Comment by xoviat** [comment]
@mpdavis 

Should be done now. 

With patch applied, requirement is cryptography except on pypy, and runtime behavior is that fallback to pycrypto will occur if cryptography cannot be imported.

File: jose/jwk.py [hunk 1]
```diff
@@ -209,43 +212,65 @@ def __init__(self, key, algorithm):
...
             raise JWKError(e)
 
     def verify(self, msg, sig):
         try:
-            return PKCS1_v1_5.new(self.prepared_key).verify(self.hash_alg.new(msg), sig)
+            self.prepared_key.verify(
+                sig,
+                msg,
+                padding.PKCS1v15(),
+                self.hash_alg(),
```
Comment (xoviat) [inline jose/jwk.py]: One thing to note is that I'm not completely sure what happens when the hash_alg is not one of the SHA* hashes. Maybe some feedback on that?

**Comment by mpdavis** [comment]
I'm not of fan of trying to fake the cryptography api with PyCrypto.  There is no requirement that each hashing algorithm needs to be implemented in a single Key class.  I would make a new class that implements the Key interface and is implemented with the cryptography library. 

That key class could then be registered over top of the PyCrypto class (or set as the default, allowing the PyCrypto class to overwrite the cryptography version).

Doing so would keep all of that code separate and much easier to understand.

**Comment by xoviat** [comment]
> I'm not of fan of trying to fake the cryptography api with PyCrypto. There is no requirement that each hashing algorithm needs to be implemented in a single Key class. I would make a new class that implements the Key interface and is implemented with the cryptography library.

Much of the code is shared between classes, which means that we would have code duplication with that option. 

> Doing so would keep all of that code separate and much easier to understand.

Not necessarily. Cryptography is the de facto cryptographic library in Python which means its functions are extensively documented. PyCrypto is significantly less documented which means that its functions are more unclear. I can add docstrings to the compat code to ease maintainability.

**Comment by mpdavis** [comment]
We have a system in place for swapping out key implementations.  I don't see how ignoring that system and doing it in a one-off manner is going to be more maintainable, despite some code duplication.  Code that does what the maintainers expect it to is important.

**Comment by xoviat** [comment]
Okay, I'll fold.

**Comment by xoviat** [comment]
Alright, should be done now. I had some trouble with tox and some ancient testing requirements, so I took it out.

File: jose/cryptphy.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
+import six
+
+from cryptography.hazmat.backends import default_backend
```
Comment (bjmc) [inline jose/cryptphy.py]: Are all the `hazmat` imports necessary because cryptography doesn't (yet?) expose these algorithms in their 'public' layer?

**Comment by bjmc** [comment]
:+1: I strongly endorse getting away from insecure, unmaintained `pycrypto`, and `cryptography` does seem to have the most momentum behind it.

**Comment by bjmc** [comment]
Is it worth putting `pycrypto` and `cryptphy` into a package like `jose._crypto_backends.cryptphy`, instead of having them both loose at the top level?

**Comment by xoviat** [comment]
@bjmc To be perfectly honest, I am no longer going to maintain this PR because I have convinced other projects to remove python-jose as a dependency. Feel free to fork this if you wish, though.

**Comment by mpdavis** [comment]
Fixed in https://github.com/mpdavis/python-jose/pull/49

**Pull request closed without merging**